✨ Add QC Metrics Data to Aligned Reads Files (#506)

mistryrn opened the pull request:
Fetch corresponding Aligned Reads QC file data from RDPC gateway to add `metrics` data to Aligned Reads files

Commits:
* 📦️ Add `graphql-request` Library
* 🔧 Add Data Center Gateway URL to Config
* ✨ Add QC Metrics Data to Aligned Reads Files
* 🛂 Create Ego Score Proxy

mistryrn commented on src/external/dataCenterGateway.ts:
```diff
@@ -0,0 +1,53 @@
+  const url = config.datacenter.gatewayUrl;
+  const graphQLClient = new GraphQLClient(url, {
+    headers: {
+      authorization: `Bearer ${config.auth.jwtKey}`,
```
@joneubank one thing I was unsure of here was whether or not this `config.auth.jwtKey` would have the correct auth for RDPC. My local environment has auth turned off for files-service, so when I was testing locally I just hardcoded my own jwt in.

mistryrn commented on src/config.ts:
```diff
@@ -166,6 +170,11 @@ const buildAppConfig = async (secrets: any): Promise<AppConfig> => {
       writeScope: `${policy}.WRITE`,
       readScope: `${policy}.READ`,
+      egoRootRest: process.env.EGO_ROOT_REST || 'http://localhost:8081',
+      egoScoreProxyClientId:
```
I followed the same pattern as a few other Vault-protected secrets for this. Haven't tested the Vault stuff yet, just been setting these in my .env

joneubank replied:
same change for the ENV file variables as well...

mistryrn commented on src/services/ego.ts:
```diff
@@ -0,0 +1,114 @@
+  return (authResponse as EgoAccessToken).access_token;
+};
+
+export const createScoreAuthClient = async () => {
```
I copied only the necessary pieces from https://github.com/icgc-argo/platform-api/blob/develop/src/services/ego/scoreAuthClient.ts and https://github.com/icgc-argo/platform-api/blob/develop/src/services/ego/index.ts for this

joneubank replied:
for naming consistency, we aren't getting auth for Score here, just getting Ego auth in general. lets rename to `createAuthClient`

mistryrn commented on .env.example:
```diff
@@ -81,6 +84,7 @@ DC_REGISTRY_URL=https://dc.dev.argo.cancercollaboratory.org/
 DC_ID=song.collab
 DC_URL=https://song.rdpc.cancercollaboratory.org/
 DC_BATCH_SIZE=50
+DC_GATEWAY_URL=https://api.rdpc.cancercollaboratory.org/graphql
```
Followed the same pattern as for the `DC_URL` variable in the .env.example here, giving the example of the prod endpoint.

joneubank replied:
lets change this to the QA URL as the provided default for devs: https://api.rdpc-qa.cancercollaboratory.org/graphql

joneubank commented on src/config.ts:
```diff
@@ -54,6 +54,9 @@ export interface AppConfig {
     writeScope: string;
     readScope: string;
+    egoRootRest: string;
+    egoScoreProxyClientId: string;
+    egoScoreProxyClientSecret: string;
```
lets drop the `scoreProxy` part of the name, now that we know it doesnt use the score-proxy permissions.

joneubank reviewed (changes requested):
Looks great, thanks for the updates. Two requests for change:
1 - naming of auth stuff we don't want the name "Score" since that isnt what we need auth for. Let's keep it generic as suggested in comments
2 - GQL Query file location - awesome job keeping this separate from the JS logic. Can you make a directory for `external/dataCenterGateway` and then move the `/gql` directory inside there like `external/dataCenterGateway/gql/QUERY_EXAMPLE.gql`, since that is the service the gql belongs to. we can then move `dataCenterGateway.ts` to `external/dataCenterGateway/index.ts` and it won't even impact your imports.

mistryrn commented:
Thanks for the feedback, @joneubank ! All really good ideas. I have renamed things to be more generic, and refactored to move the data center gateway code into its own directory within src/external
